packages/react-hooks/src/useRefEffect.ts

```diff
@@ -35,31 +35,34 @@ export function useRefEffect<T>(callback: (value: T) => (() => void) | void, ini
     callback: (value: T) => (() => void) | void;
     cleanup?: (() => void) | void;
   };
-
-  const data = React.useRef<RefData>({
-    ref: Object.assign(
-      (value: T | null) => {
-        if (data.ref.current !== value) {
-          if (data.cleanup) {
-            data.cleanup();
-            data.cleanup = undefined;
+  const cleanup = React.useRef<RefData['cleanup']>();
+  const [ref] = React.useState(() => ({
```

Comment:
I don't think you need both the `useRef` and `useState`. The `cleanup` value could be a property on the `ref` object instead of a separate `useRef`.

And then replace `cleanup.current` with `ref.cleanup`

```suggestion
  const [ref] = React.useState(() => ({
    cleanup: undefined as (() => void) | void | undefined,
```